relay-server/src/actors/upstream.rs

```diff
@@ -284,43 +319,114 @@ impl UpstreamRelay {
     pub fn new(config: Arc<Config>) -> Self {
         UpstreamRelay {
             backoff: RetryBackoff::new(config.http_max_retry_interval()),
+            network_status_backoff: RetryBackoff::new(config.http_max_retry_interval()),
             max_inflight_requests: config.max_concurrent_requests(),
             auth_state: AuthState::Unknown,
             num_inflight_requests: 0,
             high_prio_requests: VecDeque::new(),
             low_prio_requests: VecDeque::new(),
             first_error: None,
+            authentication_denied: false,
+            is_waiting_to_authenticate: false,
             config,
         }
     }
 
+    /// Predicate, checks if a Relay performs authentication
+    fn authenticates(&self) -> bool {
```

Comment:
_nit_: This function name suggests to me that it says whether Relay is _currently authenticating_, but it in fact says whether Relay _should in principle authenticate_.

```suggestion
    fn should_authenticate(&self) -> bool {
```

Same for `re_authenticate`, I would propose `should_reauthenticate`.

Reply:
ok, will do